codersdk/toolsdk/toolsdk.go

```diff
@@ -0,0 +1,1241 @@
+package toolsdk
+
+import (
+	"archive/tar"
+	"context"
+	"io"
+
+	"github.com/google/uuid"
+	"github.com/kylecarbs/aisdk-go"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+)
+
+// HandlerFunc is a function that handles a tool call.
+type HandlerFunc[T any] func(ctx context.Context, args map[string]any) (T, error)
+
+type Tool[T any] struct {
+	aisdk.Tool
+	Handler HandlerFunc[T]
+}
+
+// Generic returns a Tool[any] that can be used to call the tool.
+func (t Tool[T]) Generic() Tool[any] {
+	return Tool[any]{
+		Tool: t.Tool,
+		Handler: func(ctx context.Context, args map[string]any) (any, error) {
+			return t.Handler(ctx, args)
+		},
+	}
+}
+
+var (
+	All = []Tool[any]{
+		ReportTask.Generic(),
+		CreateWorkspace.Generic(),
+		ListWorkspaces.Generic(),
+		ListTemplates.Generic(),
+		ListTemplateVersionParameters.Generic(),
+		Whoami.Generic(),
```

Comment:
Do you think the language models will interpret this better than GetAuthenticatedUser? I'm concerned that this tool sounds more existential - like trying to define the who of I am heh

Reply:
I've anecdotally seen LMs use this tool unbidden, I'm going for the similarity to the `whoami` command. But can rename if you prefer 👍